store/rootmulti/store.go

```diff
@@ -294,22 +292,59 @@ func (rs *Store) LastCommitID() types.CommitID {
 
 // Commit implements Committer/CommitStore.
 func (rs *Store) Commit() types.CommitID {
-
-	// Commit stores.
-	version := rs.lastCommitInfo.Version + 1
+	previousHeight := rs.lastCommitInfo.Version
+	version := previousHeight + 1
 	rs.lastCommitInfo = commitStores(version, rs.stores)
 
-	// write CommitInfo to disk only if this version was flushed to disk
-	if rs.pruningOpts.FlushVersion(version) {
-		flushCommitInfo(rs.db, version, rs.lastCommitInfo)
+	// Determine if the current height needs to be added to the list of heights to
+	// be pruned.
+	if int64(rs.pruningOpts.KeepRecent) < previousHeight {
+		pruneHeight := previousHeight - int64(rs.pruningOpts.KeepRecent)
+		// We consider this height to be pruned iff:
+		//
+		// - KeepEvery is zero as that means that all heights should be pruned.
+		// - KeepEvery % (height - KeepRecent) != 0 as that means the height is not
+		// a 'snapshot' height.
+		if rs.pruningOpts.KeepEvery == 0 || pruneHeight%int64(rs.pruningOpts.KeepEvery) != 0 {
+			rs.pruneHeights = append(rs.pruneHeights, pruneHeight)
+		}
+	}
+
+	// batch prune if the current height is a pruning interval height
+	if rs.pruningOpts.Interval > 0 && version%int64(rs.pruningOpts.Interval) == 0 {
+		rs.pruneStores()
 	}
 
-	// Prepare for next version.
-	commitID := types.CommitID{
+	flushCommitInfo(rs.db, version, rs.lastCommitInfo)
+
+	return types.CommitID{
 		Version: version,
 		Hash:    rs.lastCommitInfo.Hash(),
 	}
-	return commitID
+}
+
+// pruneStores will batch delete a list of heights from each mounted sub-store.
+// Afterwards, pruneHeights is reset.
+func (rs *Store) pruneStores() {
+	if len(rs.pruneHeights) == 0 {
+		return
+	}
+
+	for key, store := range rs.stores {
+		if store.GetStoreType() == types.StoreTypeIAVL {
+			// If the store is wrapped with an inter-block cache, we must first unwrap
+			// it to get the underlying IAVL store.
+			store = rs.GetCommitKVStore(key)
+
+			if err := store.(*iavl.Store).DeleteVersions(rs.pruneHeights...); err != nil {
+				if errCause := errors.Cause(err); errCause != nil && errCause != iavltree.ErrVersionDoesNotExist {
+					panic(err)
+				}
+			}
+		}
+	}
+
+	rs.pruneHeights = make([]int64, 0)
```

Comment:
why not `rs.pruneHeights = nil`?

Reply:
Because we (potentially) append on the next block.

Reply:
It is possible to append to a nil slice, isn't it?